core/src/workflow/machines/upsert_search_attributes_state_machine.rs

```diff
@@ -1,40 +1,239 @@
 
-impl UpsertCommandCreated {
-    pub(super) fn on_upsert_workflow_search_attributes(
-        self,
-    ) -> UpsertSearchAttributesMachineTransition<UpsertCommandRecorded> {
-        unimplemented!()
+    fn try_from(e: HistoryEvent) -> Result<Self, Self::Error> {
+        match e.event_type() {
+            EventType::UpsertWorkflowSearchAttributes => {
+                Ok(UpsertSearchAttributesMachineEvents::CommandRecorded)
+            }
+            _ => Err(Self::Error::Nondeterminism(format!(
+                "UpsertWorkflowSearchAttributesMachine does not handle {e}"
+            ))),
+        }
     }
 }
 
-#[derive(Default, Clone)]
-pub(super) struct UpsertCommandRecorded {}
+// Converts generic state machine command type CommandType::UpsertWorkflowSearchAttributes into
+// the UpsertSearchAttributesMachine-specific event
+impl TryFrom<CommandType> for UpsertSearchAttributesMachineEvents {
+    type Error = WFMachinesError;
+
+    fn try_from(c: CommandType) -> Result<Self, Self::Error> {
+        match c {
+            CommandType::UpsertWorkflowSearchAttributes => {
+                Ok(UpsertSearchAttributesMachineEvents::CommandScheduled)
+            }
+            _ => Err(Self::Error::Nondeterminism(format!(
+                "UpsertWorkflowSearchAttributesMachine does not handle command type {c:?}"
+            ))),
+        }
+    }
+}
+
+// There is no Command/Reponse associated with this transition
+impl From<CommandIssued> for Done {
+    fn from(_: CommandIssued) -> Self {
+        Self {}
+    }
+}
+
+// There is no Command/Reponse associated with this transition
+impl From<Created> for CommandIssued {
+    fn from(_: Created) -> Self {
+        Self {}
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use temporal_sdk::{WfContext, WorkflowFunction};
+    use temporal_sdk_core_protos::coresdk::common::Payload;
+
+    use crate::{replay::TestHistoryBuilder, workflow::managed_wf::ManagedWFFunc};
+
+    use super::super::OnEventWrapper;
+    use super::*;
+
+    #[tokio::test]
+    async fn upsert_search_attrs_from_workflow() {
+        let mut t = TestHistoryBuilder::default();
+        t.add_by_type(EventType::WorkflowExecutionStarted);
+        t.add_full_wf_task();
+        t.add_workflow_execution_completed();
+
+        let wff = WorkflowFunction::new(|ctx: WfContext| async move {
+            ctx.upsert_search_attributes([(
+                String::from("foo"),
+                Payload {
+                    metadata: Default::default(),
+                    data: vec![],
+                },
+            )]);
+            Ok(().into())
+        });
+        let mut wfm = ManagedWFFunc::new(t, wff, vec![]);
+
+        wfm.get_next_activation().await.unwrap();
+        let commands = wfm.get_server_commands().commands;
+        assert!(!commands.is_empty());
+        assert_eq!(
+            commands[0].command_type,
+            CommandType::UpsertWorkflowSearchAttributes as i32
+        );
```

Comment:
Probably worth asserting the name/payload come out as expected too. I didn't do a great job of this everywhere in existing tests so blame me for the bad example. A lot of them are handled at the integ test layer where it's a bit easier, but in this case the workflow never gets any "result" from the command, so it's useful to ensure it's sent with the right details.